signal-desktop: fix darwin build

dwt opened the pull request:
## Things done

Getting the darwin port started, so that we can deprecate the binary package (if we want). For details see #436458

## TODOs
- [x] `nix build .#signal-desktop.libsignal-node`
- [x] `nix build .#signal-desktop.signal-sqlcipher`
- [x] `nix build .#signal-desktop.sticker-creator`
- [x] `nix develop .#signal-desktop.webrtc`
- [x] `nix build .#signal-desktop.ringrtc`
- [x] `nix build .#signal-desktop`

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [x] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dwt committed:
signal-desktop: fix darwin build of signal-sqlcipher

dwt commented:
@yuyuyureka It seems you helped get the build on linux started, do you remember where you got that `webrtc-sources.json` from? I wasn't yet able to determine that from the git history.

FlameFlag commented on pkgs/by-name/si/signal-desktop/signal-sqlcipher.nix:
```diff
@@ -71,6 +77,6 @@ stdenv.mkDerivation (finalAttrs: {
       bsd3
     ];
-    platforms = lib.platforms.linux;
+    platforms = lib.platforms.linux ++ lib.platforms.darwin;
```
It should be `lib.platforms.unix`

dwt replied:
@FlameFlag `unix` contains stuff like Cygwin, FreeBSD, Solaris, NetBSD, OpenBSD and Redox. I have not tested on any of these and am pretty sure there will be more work required to get this building on any of those platforms.

afh commented:
@dwt please find below a rough/draft patch with which `nix build .#signal-desktop.webrtc` builds without error. Hopefully that's a helpful kickstart.

[signal-desktop.webrtc-darwin.patch](https://github.com/user-attachments/files/22043023/signal-desktop.webrtc-darwin.patch)

afh committed:
signal-webrtc: fix darwin build

Add Darwin-specific toolchain and compiler runtime configuration so Signal's vendored WebRTC can build on macOS while preserving existing Linux behavior.

Co-authored-by: Angel J <78835633+iamanaws@users.noreply.github.com>

iamanaws committed:
signal-webrtc: fix darwin build with nix clang toolchain

Update the Darwin WebRTC setup to match current Chromium GN files, route
compiler-rt lookup to Nix-provided runtime libraries, and use a clang+llvm
bintools base path so llvm-ar is available. Also disable Chromium-only clang
plugins/modules and use the host-arch mac toolchain target to keep the
Signal ringrtc build working on macOS.

iamanaws committed:
signal-desktop: add Darwin packaging support

Build Signal Desktop with macOS electron-builder settings and disable
upstream signing/release hooks for nix builds. Install and run the packaged
Signal.app on Darwin, and rebuild fs-xattr explicitly so preload native
bindings are present at runtime.

iamanaws commented:
I got `signal-desktop.ringrtc` and `signal-desktop` working! 

I made a few changes over this PR + the webrtc.patch

You can get the changes from:
https://github.com/iamanaws/nixpkgs/tree/pkgs/signal-desktop

This has been tested for basic functionality on aarch64, but polish and further testing may be required to ensure everything works as expected

iamanaws commented:
@dwt at first I didn't take the time to properly fix them, but I gave it a second look and got rid of them.

[pkgs/signal-desktop](https://github.com/iamanaws/nixpkgs/tree/pkgs/signal-desktop)

---

I'm okay with you pulling the commits to this PR

marcin-serwin commented on pkgs/by-name/si/signal-desktop/webrtc.nix:
```diff
@@ -64,6 +93,17 @@ stdenv.mkDerivation (finalAttrs: {
+  ''
+  + lib.optionalString stdenv.isDarwin ''
+    substituteInPlace build/config/clang/BUILD.gn \
+      --replace-fail '_clang_lib_dir = "$clang_base_path/lib/clang/$clang_version/lib"' '_clang_lib_dir = "${llvmPackages.compiler-rt}/lib"'
```
NIT: it's a bit too long for my taste
```suggestion
      --replace-fail '_clang_lib_dir = "$clang_base_path/lib/clang/$clang_version/lib"' \
                     '_clang_lib_dir = "${llvmPackages.compiler-rt}/lib"'
```

marcin-serwin commented on pkgs/by-name/si/signal-desktop/package.nix:
```diff
@@ -200,6 +217,30 @@ stdenv.mkDerivation (finalAttrs: {
+        .afterAllArtifactBuild
+      ) |
+      .build.mac |= del(.sign)
+    ' package.json > package.json.tmp
```
Isn't this already taken care of in the `postPatch`?

marcin-serwin reviewed (changes requested):
`aarch64-linux` is still working, can't test the darwin build.

This will slightly conflict with https://github.com/NixOS/nixpkgs/pull/487711, long term it would be nice to move some of the darwin specific code into the hook.

I would appreciate if someone interested in the darwin build joined the `signal-desktop` maintainers, otherwise it will probably bitrot after few updates.

iamanaws commented:
@dwt I end up creating a separate PR to be able to follow up with this: https://github.com/NixOS/nixpkgs/pull/491588
Thanks for the help!

iamanaws commented on pkgs/by-name/si/signal-desktop/package.nix:
```diff
@@ -200,6 +217,30 @@ stdenv.mkDerivation (finalAttrs: {
     cp -r ${signal-sqlcipher} node_modules/@signalapp/sqlcipher
+  ''
+  + lib.optionalString stdenv.hostPlatform.isDarwin ''
+    # fs-xattr is required at runtime by preload.wrapper.js,
```
yeah, no problem if it does not break anything on the Linux side

iamanaws commented on pkgs/by-name/si/signal-desktop/package.nix:
```diff
@@ -212,19 +253,31 @@ stdenv.mkDerivation (finalAttrs: {
     mkdir -p $out/share/
-    cp -r dist/*-unpacked/resources $out/share/signal-desktop
+    mkdir -p $out/share/signal-desktop
+    cp -r dist/*-unpacked/resources/* $out/share/signal-desktop
```
unintentional

iamanaws commented on pkgs/by-name/si/signal-desktop/webrtc.nix:
```diff
@@ -12,9 +13,20 @@
+    name = "signal-webrtc-llvm-cc-and-bintools";
+    paths = [
+      buildPackages.rustc.llvmPackages.llvm
+      buildPackages.rustc.llvmPackages.stdenv.cc
```
fixed

eclairevoyant reviewed (approved):
<img width="488" height="485" alt="signal screenshot" src="https://github.com/user-attachments/assets/7da99246-2904-47de-85e9-45ed1a4a96f2" />


Builds and runs, messages loaded as well
